File: apps/api/src/infrastructure/repositories/activity-log-repository.ts
import { desc, eq } from 'drizzle-orm';
import { activityLogs, adminUsers } from '../../db/schema.js';
import type { AppDb } from '../database/app-db.js';

export class ActivityLogRepository {
	constructor(private readonly db: AppDb) {}

	async insert(input: {
		adminUserId: number | null;
		action: string;
		entityType: string;
		entityId?: string | null;
		details?: Record<string, unknown> | null;
	}) {
		await this.db.insert(activityLogs).values({
			adminUserId: input.adminUserId,
			action: input.action,
			entityType: input.entityType,
			entityId: input.entityId ?? null,
			details: input.details ?? null
		});
	}

	listWithAdminEmail(limit: number) {
		return this.db
			.select({
				id: activityLogs.id,
				action: activityLogs.action,
				entityType: activityLogs.entityType,
				entityId: activityLogs.entityId,
				details: activityLogs.details,
				createdAt: activityLogs.createdAt,
				adminEmail: adminUsers.email
			})
			.from(activityLogs)
			.leftJoin(adminUsers, eq(activityLogs.adminUserId, adminUsers.id))
			.orderBy(desc(activityLogs.createdAt))
			.limit(limit);
	}
}

File: apps/python-api/src/yourmedportal_api/infrastructure/repositories/activity_log_repository.py
from __future__ import annotations

import json
from typing import Any

from sqlalchemy import desc, select
from sqlalchemy.orm import Session

from yourmedportal_api.db.models import ActivityLog, AdminUser


class ActivityLogRepository:
    def __init__(self, session: Session) -> None:
        self._s = session

    def insert(
        self,
        admin_user_id: int | None,
        action: str,
        entity_type: str,
        entity_id: str | None = None,
        details: dict[str, Any] | None = None,
    ) -> None:
        log = ActivityLog(
            admin_user_id=admin_user_id,
            action=action,
            entity_type=entity_type,
            entity_id=entity_id,
            details=details,
        )
        self._s.add(log)

    def list_with_admin_email(self, limit: int) -> list[dict[str, Any]]:
        q = (
            select(
                ActivityLog.id,
                ActivityLog.action,
                ActivityLog.entity_type,
                ActivityLog.entity_id,
                ActivityLog.details,
                ActivityLog.created_at,
                AdminUser.email,
            )
            .select_from(ActivityLog)
            .outerjoin(AdminUser, ActivityLog.admin_user_id == AdminUser.id)
            .order_by(desc(ActivityLog.created_at))
            .limit(limit)
        )
        rows = self._s.execute(q).all()
        out: list[dict[str, Any]] = []
        for r in rows:
            det = r[4]
            if det is not None and not isinstance(det, dict):
                if isinstance(det, str):
                    try:
                        det = json.loads(det)
                    except Exception:
                        det = None
            out.append(
                {
                    "id": r[0],
                    "action": r[1],
                    "entityType": r[2],
                    "entityId": r[3],
                    "details": det,
                    "createdAt": r[5],
                    "adminEmail": r[6],
                }
            )
        return out
